privacyidea/api/validate.py

```diff
@@ -539,3 +540,56 @@ def trigger_challenge():
 
     return send_result(result_obj, details=details)
 
+
+@validate_blueprint.route('/polltransaction', methods=['GET'])
+@prepolicy(mangle, request=request)
+@CheckSubscription(request)
+@prepolicy(api_key_required, request=request)
+def poll_transaction():
+    """
+    Given a mandatory transaction ID, check if any non-expired challenge for this transaction ID
+    has been answered. In this case, return true. If this is not the case, return false.
+    This endpoint also returns false if no challenge with the given transaction ID exists.
+
+    This is mostly useful for out-of-band tokens that should poll this endpoint
+    to determine when to send an authentication request to ``/validate/check``.
+
+    :jsonparam transaction_id: a transaction ID
+    """
+    transaction_id = getParam(request.all_data, "transaction_id", required)
+    # Fetch a list of challenges with the given transaction ID
+    # and determine whether it contains at least one non-expired answered challenge.
+    matching_challenges = get_challenges(transaction_id=transaction_id)
```

Comment:
Wouldn't it be nicer to enhance get_challenges like

    get_challenges(transaction_id=transaction_id, otp_valid=True)

so that we only have one call?

Reply:
This would be a possibility! But then we would have two SQL queries -- one for *all* challenges of the transaction, and another one for all challenges of the transaction with ``otp_valid=True``. So I did this differently here, in order to have only one SQL query.

But if we want to change this to not log the unanswered challenges (see discussion below), I agree your solution seems better.